### ISSUE TITLE
Training Method & Dataset Structure

### ISSUE THREAD

**Issue by samyak112** [samyak112/Scout#3]
### **What We're Building**

We're creating a model that learns **directional information gain** between sentences. Think of it like this:

Imagine you're reading a document sentence by sentence. After reading sentence A, you encounter sentence B. The model answers: **"Does B add functional value to my understanding of A?"**

This is NOT about:

- Are A and B similar?
- Are A and B about the same topic?
- Do A and B share keywords?

This IS about:

- Does B resolve something incomplete in A?
- Does B explain why A is true?
- Does B provide the next step after A?
- Does B elaborate on A's meaning?

---

### **The Training Data Structure**

Each training example consists of:

**1. A Batch of Sentences** (12-15 sentences)

- Think of this as a "mini-document"
- Contains multiple independent narratives/threads
- Like a book chapter that discusses 2-3 different topics

**2. A Target Matrix** (N×N binary grid)

- For each pair of sentences (A, B), we label:
    - `1` = "Given A, B adds value"
    - `0` = "Given A, B adds nothing"
- This is **directional**: A→B might be 1, but B→A might be 0

**3. Domain Labels** (metadata)

- Which domain each sentence belongs to
- Which thread within that domain
- What role it plays (problem, solution, concept, etc.)

---

### **The Two-Domain Batch Architecture**

Each batch is structured like a newspaper page with two unrelated articles:

```
┌─────────────────────────────────────┐
│  BATCH (14 sentences)               │
│                                     │
│  ┌─────────────────────────────┐   │
│  │ DOMAIN A: Technology        │   │
│  │                             │   │
│  │ Thread 1: Python Crash      │   │
│  │   [0] Problem               │   │
│  │   [1] Diagnosis             │   │
│  │   [2] Solution              │   │
│  │                             │   │
│  │ Thread 2: ML Concepts       │   │
│  │   [3] Concept A             │   │
│  │   [4] Concept B             │   │
│  │                             │   │
│  │ Hard Negatives              │   │
│  │   [5] Tech trivia           │   │
│  │   [6] Generic tech fact     │   │
│  └─────────────────────────────┘   │
│                                     │
│  ┌─────────────────────────────┐   │
│  │ DOMAIN B: Finance           │   │
│  │                             │   │
│  │ Thread 3: Portfolio Loss    │   │
│  │   [7] Problem               │   │
│  │   [8] Diagnosis             │   │
│  │   [9] Solution              │   │
│  │                             │   │
│  │ Thread 4: Investment Theory │   │
│  │   [10] Concept A            │   │
│  │   [11] Concept B            │   │
│  │                             │   │
│  │ Hard Negatives              │   │
│  │   [12] Finance trivia       │   │
│  │   [13] Generic finance fact │   │
│  └─────────────────────────────┘   │
└─────────────────────────────────────┘
```

---

#### **Signal 1: Inter-Domain Separation (Easy)**

- Domain A sentences ↔ Domain B sentences = ZERO
- "Python crash" has nothing to do with "Portfolio loss"
- Even if they share words, different contexts = no connection

**Purpose:** Teach the model not to hallucinate connections between unrelated topics

#### **Signal 2: Intra-Domain Selectivity (Hard)**

- Within Domain A:
    - Thread 1 (Python crash) ↔ Thread 2 (ML concepts) = ZERO
    - Both are "about Python" but serve different purposes
    - Model must learn: same topic ≠ automatic relevance

**Purpose:** Teach the model that functional relevance requires more than keyword overlap

#### **Signal 3: Within-Thread Directional Chains (Core Learning)**

- Within Thread 1 (Python crash):
    - [0] Problem → [1] Diagnosis = 1 (adds value)
    - [1] Diagnosis → [2] Solution = 1 (adds value)
    - [2] Solution → [0] Problem = 0 (no added value)

**Purpose:** Teach the model directional information flow

#### **Signal 4: Hard Negative Rejection (Critical)**

- [5] Tech trivia → ALL threads = ZERO
- Shares domain but adds no functional value
- Model must learn: being "on topic" isn't enough

**Purpose:** Teach the model to distinguish informative from uninformative content

---

#### **Step 1: Batch Presentation**

```
Input: 14 sentences [S0, S1, ..., S13]
Output: 14×14 matrix of predictions
```

The model sees all sentences and predicts for each pair (Si, Sj):

- P(Sj adds value | you just read Si)

#### **Step 2: Loss Calculation**

```
For each cell (i,j) in the matrix:
  loss += binary_cross_entropy(predicted[i][j], target[i][j])
```

The model is penalized when:

- It predicts 1 but target is 0 (false positive - hallucinating connections)
- It predicts 0 but target is 1 (false negative - missing connections)

#### **Step 3: Learning Signals**

The model adjusts its weights to:

1. **Separate domains** (inter-domain = 0)
2. **Separate threads within domain** (intra-domain selectivity)
3. **Connect within threads** (directional chains)
4. **Reject hard negatives** (same domain, no value)

---

#### **Phase 1: Foundation (100-200 batches)**

Goal: Teach basic patterns

Distribution:

- 40% Same domain, different threads (hardest)
- 30% Different domains, clean separation (easier)
- 20% Within-thread directional chains
- 10% All-zero batches (null test)

#### **Phase 2: Complexity (500-1000 batches)**

Goal: Add nuance

Add:

- Polysemy traps (same word, different meanings)
- Graduated relevance (weak vs strong connections)
- Multi-hop chains (A→B→C)
- Contradictions and negations

#### **Phase 3: Real-World (2000+ batches)**

Goal: Handle messy data

Add:

- Variable domain proportions (70/30, 80/20)
- Cross-domain rare connections
- Long-form content (20-30 sentences)
- Domain-specific jargon

**Comment by samyak112** [comment]
<img width="856" height="655" alt="Image" src="https://github.com/user-attachments/assets/4997b26b-9ad1-49b8-8a7d-93a5cb9b470e" />


<img width="845" height="582" alt="Image" src="https://github.com/user-attachments/assets/6fca1286-4103-404d-a9d9-d226a3fb706a" />

**Comment by samyak112** [comment]
Here is what I need to do:

Test 1 - Sanity Check:
I'll take a single example pair and train on it for 100 steps. If the loss goes to 0.00, my code works. If it doesn't, I have a bug.

Test 2 - Overfit Check:
I'll create 10-20 diverse examples with clear asymmetric relationships (problems→solutions, questions→answers, causes→effects). I'll train on all of them until loss hits ~0, then run inference on those same examples. If my predictions match my labels perfectly, my architecture can learn asymmetry. If not, my architecture has a fundamental limitation.

Test 3 - Generalization Check:
I'll create 100-200 examples and split them 80/20. I'll train only on the 80%, then test on the unseen 20%. If I get >70% accuracy on the test set, my model actually learns patterns instead of just memorizing. This proves it could work in the real world

**Comment by samyak112** [comment]
### How do I make the dataset?

I initially thought that i will use LLM to make the data set basically i will input this training structure in LLM and expect to return some array , BUTTT I dont think its going to be that easy

If i asked an LLM to do this there are chances that LLM will hallucinate with the numbers of my matrix i think its better that 
I force the LLM to generate the **metadata**, and build the matrix deterministically in Python.

**The Protocol:**

1. **Prompt 1 (Thread A):** "Write a 3-sentence functional chain about [Topic] with roles: [Problem], [Diagnosis], [Solution]."
    
2. **Prompt 2 (Thread B):** "Write a 2-sentence concept definition chain about [Unrelated Topic]."
    
3. **Prompt 3 (Hard Negatives):** "Write 2 sentences that use the _same terminology_ as Thread A but are functionally useless trivia."
    

**The Validation Logic (Python-side):**

I implicitly know the relationships because I defined the structure _before_ generation.

- $T_A[0] \to T_A[1]$ is **1** (by definition).
    
- $T_A \to T_B$ is **0** (by definition).
    
- $Negative \to T_A$ is **0** (by definition).

**Comment by samyak112** [comment]
### Assymetry cant be just one type

Up until now i was planning to keep only one sort of assymetry which is problem diagnose solution, but that is not a "type" of assymetry per se its more of an example of assymetry , so i tried to see that linguistically what sorts of assymetries are possible and got this.

### 1. Temporal Asymmetry (The "Iconicity" Principle)

This is the most primal form. Language often mimics the physical flow of time. If Event A happens before Event B in reality, we prefer to say them in that order. Violating this creates confusion or requires extra grammatical "glue" (like "before" or "after").

- **The Principle:** The order of clauses mirrors the order of events.
    
- **Example:** "He opened the door and walked in."
    
    - _Asymmetric:_ You cannot say "He walked in and opened the door" without changing the physical reality of what happened.
        
- **Why it's directional:** It relies on the arrow of time.
    

### 2. Causal Asymmetry (Explanation & Result)

This is closest to your "valve and plumber" example. One idea is the _condition_ for the other. You cannot simply swap them without breaking the logical chain.

- **The Principle:** Cause must be established to understand the Effect, or the Need must be established to understand the Action.
    
- **Example:** "The window broke. The room got cold."
    
    - _Asymmetric:_ If you swap it to "The room got cold. The window broke," you imply the cold caused the break (thermal shock), changing the meaning entirely.
        
- **Your Example:** "I need to fix a valve" (Problem) $\rightarrow$ "I will call a plumber" (Solution). The second sentence has no utility without the first.
    

### 3. Epistemic Asymmetry (Given vs. New)

This is about "Information Flow." In communication, we almost always move from **Old Information** (what we both know) to **New Information** (what I am telling you). If you reverse this, the listener gets lost because you are answering a question they haven't asked yet.

- **The Principle:** You must anchor a new idea to an existing peg.
    
- **Example:**
    
    - _Correct:_ "Do you see that guy? He is my brother." (Anchor $\rightarrow$ New Info)
    - _Broken:_ "He is my brother. Do you see that guy?" (New Info $\rightarrow$ Anchor)
        
- **Why it's directional:** You cannot define a variable _after_ you have used it in a function.
    

### 4. Interactional Asymmetry (Adjacency Pairs)

This occurs in dialogue. Certain social actions compel a specific response. The first part dictates the second; they are not peers.

- **The Principle:** The first utterance creates a "slot" that forces a specific type of second utterance.
    
- **Example:** A Question demands an Answer.
    
    - _Sequence:_ "What time is it?" $\rightarrow$ "5 PM."
    - _Asymmetric:_ You cannot have the Answer precede the Question in a coherent text.
        
- **Other types:** Offer $\rightarrow$ Acceptance/Refusal; Greeting $\rightarrow$ Greeting.
    

### 5. Semantic Asymmetry (Hyponymy/Entailment)

This is a logical asymmetry found inside the definitions of words. It works one way but not the other.

- **The Principle:** Specifics entail generalities, but generalities do not entail specifics.
    
- **Example:**
    
    - _Direction A:_ "I saw a **Honda**." $\rightarrow$ This implies "I saw a **car**." (True)
    - _Direction B:_ "I saw a **car**." $\rightarrow$ This does _not_ imply "I saw a **Honda**." (False)
        
- **Why it's directional:** It is a set-subset relationship.

**Comment by samyak112** [comment]
I added weighted MSE because I realized that not all errors in my matrix matter equally. Most of the entries in my 7×7 target matrix are probably near zero, and only a smaller subset represents meaningful relationships. If I used a normal MSE, the model could minimize loss mostly by predicting zeros everywhere, because that already matches the majority of entries. It would look like the loss is going down, but it wouldn’t actually be learning the important relational structure.

By using weighted MSE, I’m telling the model: “Mistakes on important relationships cost more than mistakes on irrelevant ones.” That shifts learning pressure toward the meaningful connections instead of letting the model hide behind the many low-value entries. In practice, this helps the model focus on learning real structural signals instead of converging to a trivial solution that just predicts weak or zero relations everywhere.
